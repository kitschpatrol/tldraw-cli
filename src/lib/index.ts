export { type TldrawOpenOptions, type TldrawOpenResult, tldrawOpen } from './tldraw-open'
export { type TldrawFormat, type TldrawToImageOptions, tldrawToImage } from './tldraw-to-image'
export { default as log } from './utilities/log'
