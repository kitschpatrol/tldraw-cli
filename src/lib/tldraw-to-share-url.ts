import TldrawController from './tldraw-controller'
import { validatePathOrUrl } from './validation'

/**
 * Gets a share URL from a tldraw file or URL
 */
export async function tldrawToShareUrl(tldrPathOrUrl: string): Promise<string> {
	const validatedPathOrUrl = validatePathOrUrl(tldrPathOrUrl, {
		requireFileExistence: true,
		validFileExtensions: ['.tldr'],
		validHostnames: ['www.tldraw.com'],
	})

	if (typeof validatedPathOrUrl === 'string') {
		// It's a local file, so we need to upload it to tldraw.com
		// and then get the share link
		const tldrawController = new TldrawController('https://www.tldraw.com')
		await tldrawController.start()
		await tldrawController.loadFile(validatedPathOrUrl)
		const shareUrl = await tldrawController.getShareUrl()
		await tldrawController.close()
		return shareUrl
	}

	// It's already a tldraw.com url, just return that
	// TODO better to download and make a copy?
	return validatedPathOrUrl.href
}
