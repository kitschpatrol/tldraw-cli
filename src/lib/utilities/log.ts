/* eslint-disable perfectionist/sort-objects */
// Staying basic, always log to stderr

import chalk from 'chalk'

const isNode = process?.versions?.node !== undefined

const log = {
	verbose: false,

	// Intended for temporary logging
	log(...data: unknown[]): void {
		if (!this.verbose) return
		const levelPrefix = chalk.gray('[Log]')
		if (isNode) {
			// Log to stderr in node for ease of redirection
			console.warn(levelPrefix, ...data)
		} else {
			console.log(levelPrefix, ...data)
		}
	},
	logPrefixed(prefix: string, ...data: unknown[]): void {
		this.info(chalk.blue(`[${prefix}]`), ...data)
	},

	info(...data: unknown[]): void {
		if (!this.verbose) return
		const levelPrefix = chalk.green('[Info]')
		if (isNode) {
			// Log info to stderr in node for ease of redirection
			console.warn(levelPrefix, ...data)
		} else {
			console.info(levelPrefix, ...data)
		}
	},
	infoPrefixed(prefix: string, ...data: unknown[]): void {
		this.info(chalk.blue(`[${prefix}]`), ...data)
	},

	warn(...data: unknown[]): void {
		console.warn(chalk.yellow('[Warning]'), ...data)
	},
	warnPrefixed(prefix: string, ...data: unknown[]): void {
		this.warn(chalk.blue(`[${prefix}]`), ...data)
	},

	error(...data: unknown[]): void {
		console.error(chalk.red('[Error]'), ...data)
	},
	errorPrefixed(prefix: string, ...data: unknown[]): void {
		this.error(chalk.blue(`[${prefix}]`), ...data)
	},
}

export default log
