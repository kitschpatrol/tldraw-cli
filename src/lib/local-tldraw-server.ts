import type { Server } from 'node:http'
import type { AddressInfo } from 'node:net'
import express from 'express'
import getPort from 'get-port'
import path from 'node:path'
import { fileURLToPath } from 'node:url'
import log from './utilities/log'

export default class LocalTldrawServer {
	get href(): string {
		if (!this.server) throw new Error('Server not started')
		const { port } = this.server.address() as AddressInfo
		return `http://localhost:${port}`
	}

	private server?: Server

	constructor(private readonly tldrData?: string) {}

	close(): void {
		if (!this.server) throw new Error('Server not started')
		this.server.close()
		log.info('Stopped tldraw server')
	}

	async start() {
		// Serve local tldraw
		log.info('Starting tldraw server...')

		const scriptDirectory = path.dirname(fileURLToPath(import.meta.url))

		// Handle dev or prod relative paths, brittle
		const tldrawPath = path.join(
			scriptDirectory,
			scriptDirectory.endsWith('/src/lib')
				? '../../dist/tldraw'
				: scriptDirectory.endsWith('/dist/lib')
					? '../tldraw'
					: '../dist/tldraw',
		)

		log.info(`tldraw served from "${tldrawPath}"`)
		const app = express()
		const port = await getPort()

		// Provide the initial state data at an endpoint
		app.get('/tldr-data', (_, response) => {
			if (this.tldrData === undefined) {
				response.status(404).send('No tldr data provided')
			} else {
				response.setHeader('Content-Type', 'text/plain; charset=utf-8')
				response.send(this.tldrData)
			}
		})

		app.use(express.static(tldrawPath))

		try {
			this.server = app.listen(port)
		} catch (error) {
			log.error(error)
		}

		log.info(`tldraw hosted at "${this.href}"`)
	}
}
