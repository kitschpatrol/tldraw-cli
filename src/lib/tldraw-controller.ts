/* eslint-disable ts/no-unnecessary-condition */
/* eslint-disable unicorn/prefer-global-this */
// Note special inline IIFE import, see ./plugins/esbuild-plugin-iife.ts
import type { Browser, Page } from 'puppeteer'
import slugify from '@sindresorhus/slugify'
import * as cheerio from 'cheerio'
import { orderBy } from 'natural-orderby'
import fs from 'node:fs/promises'
import path from 'node:path'
import puppeteer from 'puppeteer'
import { base64ToString, base64ToUint8Array, stringToBase64 } from 'uint8array-extras'
import untildify from 'untildify'
import type { TldrawToImageOptions } from './tldraw-to-image'
import getImageInlineScript from './inline/get-image?iife'
import getTldrInlineScript from './inline/get-tldr?iife'
import setTldrInlineScript from './inline/set-tldr?iife'
import log from './utilities/log'

type TlPage = {
	frames: TlFrame[]
	id: string
	name: string
}
type TlFrame = { id: string; name: string }

type DownloadPlan = {
	fileSuffix: string
	frameId: string | undefined
	pageId: string
}

export default class TldrawController {
	private browser?: Browser
	private isEmpty?: boolean
	private page?: Page

	private get isLocal(): boolean {
		return this.href.startsWith('http://localhost')
	}

	constructor(private readonly href: string) {}

	async close() {
		if (!this.browser) throw new Error('Controller not started')

		// Avoid lingering Chrome processes
		// https://stackoverflow.com/a/76505750/2437832
		const pages = await this.browser.pages()
		await Promise.all(pages.map(async (page) => page.close()))

		await this.browser.close()
		log.info('Stopped controller')
	}

	// eslint-disable-next-line complexity
	async download(options: TldrawToImageOptions): Promise<string[]> {
		// Validate options at the last minute
		if (!this.page) throw new Error('Controller not started')
		if (this.isEmpty) throw new Error('Cannot export an empty document')

		// Only SVGs have style tags
		if (options.stripStyle && options.format !== 'svg')
			log.warn('--strip-style is only supported for SVG output')

		// Many flags don't make sense for TLDR
		if (options.format === 'tldr') {
			if (options.frames !== undefined)
				log.warn(
					'--frames is not supported when exporting to "tldr", ignoring flag and exporting entire sketch file',
				)

			if (options.pages !== undefined)
				log.warn(
					'--pages is not supported when exporting to "tldr", ignoring flag and exporting entire sketch file',
				)

			if (options.dark !== undefined && options.dark) {
				log.warn(
					'--dark is not supported when exporting to "tldr", ignoring flag and exporting sketch file',
				)
			}

			if (options.padding !== undefined) {
				log.warn(
					'--padding is not supported when exporting to "tldr", ignoring flag and exporting sketch file',
				)
			}

			if (options.scale !== undefined) {
				log.warn(
					'--scale is not supported when exporting to "tldr", ignoring flag and exporting sketch file',
				)
			}

			if (options.transparent !== undefined && options.transparent) {
				log.warn(
					'--transparent is not supported when exporting to "tldr", ignoring flag and exporting sketch file',
				)
			}
		}

		// Set defaults, most come from tldraw itself or the cli tools
		const {
			dark = false,
			format = 'svg',
			name: filename,
			output = './',
			padding,
			print = false,
			scale,
			stripStyle = false,
			transparent = false,
		} = this.stripUndefined(options) as TldrawToImageOptions

		// Downloading to TLDR format is a "short path"
		if (format === 'tldr') {
			await this.page.evaluate(getTldrInlineScript)
			const base64String = await this.page.evaluate(async () => window.getTldr())

			if (print) {
				const plainString = base64ToString(base64String).replaceAll('\n', '')
				return [plainString]
			}

			const outputPath = path.resolve(untildify(path.join(output, `${filename}.${format}`)))
			await fs.writeFile(outputPath, base64ToUint8Array(base64String))
			return [outputPath]
		}

		// Download to other formats, long path

		// Load the getImage function into the page
		await this.page.evaluate(getImageInlineScript)

		// Get the page / frames of sketch...
		const sketchStructure = await this.getSketchStructure()

		// Parse those down into a download plan
		// Note that this validates and throws warnings about requested pages and frames
		const downloadPlan = this.getDownloadPlans(sketchStructure, options.pages, options.frames)

		const outputAccumulator: string[] = []

		const initialPageId = await this.getCurrentPage()

		for (const download of downloadPlan) {
			// Change page if necessary
			const currentPageId = await this.getCurrentPage()
			if (download.pageId !== currentPageId) {
				log.info(`Selecting sketch page "${download.pageId}"`)

				await this.setCurrentPage(download.pageId)
			}

			// Wait until editor is available
			await this.page.waitForFunction('editor !== undefined')

			if (download.frameId === undefined) {
				await this.page.evaluate('editor.selectAll()')
			} else {
				// Select frame
				log.info(`Selecting sketch frame "${download.frameId}"`)

				// Select the frame shape
				await this.page.evaluate('editor.selectNone()')
				await this.page.evaluate(`editor.select('${download.frameId}')`)
			}

			let base64String = await this.page.evaluate(async (options) => window.getImage(options), {
				background: transparent === undefined ? undefined : !transparent,
				darkMode: dark,
				format,
				padding,
				scale,
			})

			if (stripStyle && format === 'svg') {
				base64String = stringToBase64(this.stripStyleElement(base64ToString(base64String)))
			}

			if (print) {
				if (format === 'png') {
					outputAccumulator.push(base64String)
				} else {
					// All others are plain text, without newlines
					const plainString = base64ToString(base64String).replaceAll('\n', '')
					outputAccumulator.push(plainString)
				}
			} else {
				const outputPath = path.resolve(
					untildify(path.join(output, `${filename}${download.fileSuffix}.${format}`)),
				)

				await fs.writeFile(outputPath, base64ToUint8Array(base64String))
				outputAccumulator.push(outputPath)
			}
		}

		// Restore page if necessary
		const currentPageId = await this.getCurrentPage()
		if (currentPageId !== initialPageId) {
			await this.setCurrentPage(initialPageId)
		}

		// Sort output naturally in place, addresses rare instability
		return orderBy(outputAccumulator)
	}

	async loadFile(filePath: string) {
		if (!this.page) throw new Error('Controller not started')
		if (this.isLocal)
			throw new Error(
				'File loading is only supported for remote tldraw.com instances. See tldraw-open.ts for local file loading approach.',
			)

		await this.closeMenus()
		await this.page.evaluate(`userPreferences.showFileOpenWarning.set(false);`)

		log.info(`Uploading local file to tldraw.com: ${filePath}`)
		const tldrFile = await fs.readFile(filePath, 'utf8')

		// We have to call a custom function to upload the tldr file,
		// puppeteer waitForFileChooser fileInput.accept etc. does NOT work
		await this.page.evaluate(setTldrInlineScript)
		await this.page.evaluate(`window.setTldr(${tldrFile})`)
	}

	async start() {
		// Set up Puppeteer
		log.info('Starting Puppeteer...')
		this.browser = await puppeteer.launch({
			args: this.isLocal
				? ['--no-sandbox', '--disable-web-security', '--disable-setuid-sandbox']
				: // Both contexts for now?
					['--no-sandbox', '--disable-web-security', '--disable-setuid-sandbox'],
			headless: true,
		})

		this.page = await this.browser.newPage()
		this.page.setDefaultTimeout(120_000)

		// Set up console logging passthrough
		this.page.on('console', (message) => {
			const messageType = message.type()
			const messageText = message.text()

			if (messageType === 'error') {
				log.errorPrefixed('Browser', messageText)
			} else if (messageType === 'warn') {
				log.warnPrefixed('Browser', messageText)
			} else {
				log.infoPrefixed('Browser', messageText)
			}
		})

		// Navigate to tldraw
		log.info(`Navigating to: ${this.href}`)
		await this.page.goto(this.href, { waitUntil: 'networkidle0' })

		// Wait until editor is available
		await this.page.waitForFunction('editor !== undefined')

		// Check for emptiness
		const shapeCount = (await this.page.evaluate('editor.getCurrentPageShapes().length')) as number
		this.isEmpty = shapeCount === 0
	}

	private async closeMenus(): Promise<void> {
		if (!this.page) throw new Error('Controller not started')
		await this.page.evaluate(`editor.clearOpenMenus()`)
	}

	private async getCurrentPage(): Promise<string> {
		if (!this.page) throw new Error('Controller not started')
		return (await this.page.evaluate(`editor.getCurrentPageId()`)) as string
	}

	// eslint-disable-next-line complexity
	private getDownloadPlans(
		sketch: TlPage[],
		pages?: boolean | number[] | string[],
		frames?: boolean | string[],
	): DownloadPlan[] {
		const validPages = this.validatePages(sketch, pages)
		const validFrames = this.validateFrames(sketch, validPages, frames)

		// First pick pages
		// Logic:
		// Pages undefined or false, download the first page only
		// Pages true, download all pages
		// Pages string array, download matching page ids or names, allowing slugs and prefixes
		// TODO Pages number array, download matching page indexes
		const filteredSketch =
			validPages === undefined
				? [sketch[0]]
				: typeof validPages === 'boolean' && !validPages
					? [sketch[0]]
					: sketch.filter((sketchPage) =>
							typeof validPages === 'boolean' && validPages
								? true
								: validPages.some(
										(p) =>
											slugify(p) === slugify(sketchPage.name) ||
											p.replace('page:', '') === sketchPage.id.replace('page:', ''),
									),
						)

		// Then pick frames for each page
		// Logic:
		// Frames undefined or false, download the whole page contents, so pass empty array to frames
		// Frames true, download each frame separately
		// Frames string array, download matching frame ids or names, allowing slugs and prefixes

		for (const page of filteredSketch) {
			page.frames =
				validFrames === undefined || (typeof validFrames === 'boolean' && !validFrames)
					? [] // Empty means download entire page as-is!
					: page.frames.filter((frame) =>
							typeof validFrames === 'boolean' && validFrames
								? true
								: validFrames.some(
										(f) =>
											slugify(f) === slugify(frame.name) ||
											f.replace('shape:', '') === frame.id.replace('shape:', ''),
									),
						)
		}

		// Flatten the filtered sketch into a download plan
		const downloadPlans: DownloadPlan[] = []

		// Check for page.name collisions, if we have any then include page id in the filename
		const isPageNameCollision =
			new Set(filteredSketch.map((page) => slugify(page.name))).size !== filteredSketch.length

		if (isPageNameCollision) {
			log.warn(
				'Page names are not unique, including page IDs in the output filenames to avoid collisions',
			)
		}

		for (const page of filteredSketch) {
			// Ensure page names are unique

			const pageSuffix =
				validPages === undefined || (typeof validPages === 'boolean' && !validPages)
					? undefined // No suffix unless pages are explicitly requested
					: isPageNameCollision
						? `${slugify(page.name)}-${page.id.replace('page:', '')}`
						: slugify(page.name)

			if (page.frames.length === 0) {
				downloadPlans.push({
					fileSuffix: pageSuffix ?? '',
					frameId: undefined,
					pageId: page.id,
				})
			} else {
				// Check for frame name collisions, if we have any then include frame id in the filename
				const isFrameNameCollision =
					new Set(page.frames.map((frame) => slugify(frame.name))).size !== page.frames.length

				// TODO warn once?
				if (isFrameNameCollision) {
					log.warn(
						'Frame names are not unique, including frame IDs in the output filenames to avoid collisions',
					)
				}

				for (const frame of page.frames) {
					// Frames must have been explicitly requested
					const frameSuffix = isFrameNameCollision
						? [slugify(frame.name), frame.id.replace('shape:', '')]
								.filter((value) => value !== '')
								.join('-')
						: slugify(frame.name)

					downloadPlans.push({
						fileSuffix: [pageSuffix, frameSuffix].filter((value) => value !== undefined).join('-'),
						frameId: frame.id,
						pageId: page.id,
					})
				}
			}
		}

		// Prefix suffix with dash
		for (const plan of downloadPlans) {
			if (plan.fileSuffix !== '') {
				plan.fileSuffix = `-${plan.fileSuffix}`
			}
		}

		return downloadPlans
	}

	private async getPageFrames(pageId: string): Promise<TlFrame[]> {
		if (!this.page) throw new Error('Controller not started')

		const initialPageId = await this.getCurrentPage()

		if (pageId !== initialPageId) {
			await this.setCurrentPage(pageId)
		}

		// Returns empty array if no frames
		// Note that while 'Frame' is the default name shown in the UI... it's an
		// empty string in the data until explicitly set by the user (consider i18n)
		// we could pass 'frame' here for nicer filenames
		const frames = (await this.page.evaluate(
			`editor.getCurrentPageShapes().reduce((accumulator, shape) => {
				if (shape.type === 'frame') {
					accumulator.push({ id: shape.id, name: shape.props.name === '' ? 'Frame' : shape.props.name })
				}
				return accumulator
			}, [])`,
		)) as TlFrame[]

		// Restore page if necessary
		if (pageId !== initialPageId) {
			await this.setCurrentPage(initialPageId)
		}

		return frames
	}

	private async getPages(): Promise<TlPage[]> {
		if (!this.page) throw new Error('Controller not started')

		return (await this.page.evaluate(
			`editor.getPages().map((page) => ({ id: page.id, name: page.name, frames: []}))`,
		)) as TlPage[]
	}

	// Structure
	private async getSketchStructure(): Promise<TlPage[]> {
		if (!this.page) throw new Error('Controller not started')

		const document = await this.getPages()

		for (const page of document) {
			page.frames = await this.getPageFrames(page.id)
		}

		return document
	}

	private async setCurrentPage(pageId?: string): Promise<void> {
		if (!this.page) throw new Error('Controller not started')
		await this.page.evaluate(`editor.setCurrentPage("${pageId}")`)
	}

	// Helpers
	private stripStyleElement(svg: string): string {
		const dom = cheerio.load(svg, { xmlMode: true })
		dom('style').remove()
		return dom.xml()
	}

	private stripUndefined(options: Record<string, unknown>): Record<string, unknown> {
		return Object.fromEntries(Object.entries(options).filter(([, value]) => value !== undefined))
	}

	private validateFrames(
		sketch: TlPage[],
		pages: boolean | number[] | string[] | undefined,
		frames: boolean | string[] | undefined,
	): boolean | string[] | undefined {
		if (Array.isArray(frames)) {
			const validSketch = sketch.filter(
				(page, index) =>
					pages === undefined ||
					(typeof pages === 'boolean' && !pages && index === 0) ||
					(typeof pages === 'boolean' && pages) ||
					(Array.isArray(pages) &&
						pages.some((p, index) =>
							typeof p === 'number'
								? index === p
								: slugify(p) === slugify(page.name) ||
									p.replace(/^page:/, '') === page.id.replace(/^page:/, ''),
						)),
			)

			const validFrames = []

			for (const f of frames) {
				const matchingFrames: string[] = []

				// One frame name can match multiple frame IDs across pages
				for (const page of validSketch) {
					const match = page.frames.find(
						(frame) =>
							slugify(f) === slugify(frame.name) ||
							`shape:${f.replace(/^shape:/, '')}` === frame.id,
					)

					if (match && !matchingFrames.includes(match.id)) {
						matchingFrames.push(match.id)
					}
				}

				if (matchingFrames.length === 0) {
					log.warn(`Frame "${f}" not found in sketch`)
				} else {
					validFrames.push(...matchingFrames)
				}

				if (validFrames.length === 0) {
					log.warn('None of the requested frames were found in sketch, ignoring frames option')
					return undefined
				}
			}

			return validFrames
		}

		return frames
	}

	// Deduplicates, verifies existence, and normalizes to ids of the style 'page:xxxx'
	private validatePages(
		sketch: TlPage[],
		pages: boolean | number[] | string[] | undefined,
	): boolean | string[] | undefined {
		if (Array.isArray(pages)) {
			const validPages: string[] = []
			for (const p of pages) {
				const matchingPage =
					typeof p === 'number'
						? sketch[p]
						: sketch.find(
								(page) =>
									slugify(p) === slugify(page.name) ||
									`page:${p.replace(/^page:/, '')}` === page.id,
							)
				if (matchingPage) {
					if (!validPages.includes(matchingPage.id)) {
						validPages.push(matchingPage.id)
					}
				} else {
					log.warn(`Page "${p}" not found in sketch`)
				}
			}

			if (validPages.length === 0) {
				log.warn('None of the requested pages were found in sketch, ignoring pages option')
				return undefined
			}

			return validPages
		}

		return pages
	}
}
