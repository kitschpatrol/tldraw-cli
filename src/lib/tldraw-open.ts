import { nanoid } from 'nanoid'
import { type ChildProcess } from 'node:child_process'
import fs from 'node:fs/promises'
import os from 'node:os'
import { URL } from 'node:url'
import open from 'open'
import LocalTldrawServer from './local-tldraw-server'
import { tldrawToImage } from './tldraw-to-image'
import { tldrawToShareUrl } from './tldraw-to-share-url'
import log from './utilities/log'
import { validatePathOrUrl } from './validation'

const tldrawOpenDefaultOptions = {
	location: 'remote',
}

export type TldrawOpenOptions = {
	location: 'local' | 'remote'
}

export type TldrawOpenResult = {
	browserExitPromise: Promise<ChildProcess>
	openedSketchUrl: string
}

export async function tldrawOpen(
	tldrPathOrUrl?: string,
	options?: Partial<TldrawOpenOptions>,
): Promise<TldrawOpenResult> {
	const tldrawRemoteUrl = 'https://www.tldraw.com'
	const { location } = { ...tldrawOpenDefaultOptions, ...options }

	// Set below depending on the combination of options
	let urlToOpen: string | undefined

	const validatedPathOrUrl =
		tldrPathOrUrl === undefined
			? undefined
			: validatePathOrUrl(tldrPathOrUrl, {
					requireFileExistence: true,
					validFileExtensions: ['.tldr'],
					validHostnames: ['www.tldraw.com'],
				})

	// Open empty file locally
	if (location === 'local' && validatedPathOrUrl === undefined) {
		const tldrawServer = new LocalTldrawServer()
		await tldrawServer.start()
		urlToOpen = tldrawServer.href
		log.info(`Opened blank tldraw sketch locally at "${urlToOpen}"`)
	}
	// Open local file locally
	else if (location === 'local' && typeof validatedPathOrUrl === 'string') {
		const tldrData = await fs.readFile(validatedPathOrUrl, 'utf8')
		const tldrawServer = new LocalTldrawServer(tldrData)
		await tldrawServer.start()
		urlToOpen = tldrawServer.href
		log.info(`Opened copy of tldraw sketch "${validatedPathOrUrl}" locally at "${urlToOpen}"`)
	}
	// Open remote file locally
	else if (location === 'local' && validatedPathOrUrl instanceof URL) {
		const [savedFile] = await tldrawToImage(validatedPathOrUrl.href, {
			format: 'tldr',
			name: nanoid(),
			output: os.tmpdir(),
		})
		const tldrData = await fs.readFile(savedFile, 'utf8')
		await fs.rm(savedFile, { force: true })
		const tldrawServer = new LocalTldrawServer(tldrData)
		await tldrawServer.start()
		urlToOpen = tldrawServer.href
		log.info(
			`Opened a local copy of tldraw sketch url "${validatedPathOrUrl.href}" locally at ${urlToOpen}`,
		)
	}
	// Open empty file remotely, actually just opens tldraw.com
	else if (location === 'remote' && validatedPathOrUrl === undefined) {
		urlToOpen = tldrawRemoteUrl
		log.info(`Opened tldraw.com`)
	}
	// Open local file remotely
	else if (location === 'remote' && typeof validatedPathOrUrl === 'string') {
		urlToOpen = await tldrawToShareUrl(validatedPathOrUrl)
		log.info(
			`Opened copy of local tldraw sketch "${validatedPathOrUrl}" remotely at "${urlToOpen}"`,
		)
	}
	// Open remote url remotely
	else if (location === 'remote' && validatedPathOrUrl instanceof URL) {
		urlToOpen = validatedPathOrUrl.href
		log.info(`Opened tldraw sketch url at ${urlToOpen}`)
	} else {
		// Should be unreachable
		throw new Error('Invalid tldrawOpen options')
	}

	const exitPromise = open(urlToOpen, {
		wait: true,
	})

	return {
		browserExitPromise: exitPromise,
		openedSketchUrl: urlToOpen,
	}
}
