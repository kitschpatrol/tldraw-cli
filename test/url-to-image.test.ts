// Note this tests the dist build, because of the iife inlining from esbuild
import { tldrawToImage } from '../dist/lib'
import { expectFileToBeValid } from './utilities/file'
import { nanoid } from 'nanoid'
import { mkdirSync, rmSync } from 'node:fs'
import { expect, it } from 'vitest'

const cleanUp = true
const tldrawTestUrl = 'https://www.tldraw.com/s/v2_c_9nMYBwT8UQ99RGDWfGr8H'
const tldrawTestUrlSchema2 = 'https://www.tldraw.com/s/v2_c_85CiFqkLgUaiwmed4kIa_'
const tldrawTestThreeFramesUrl = 'https://www.tldraw.com/s/v2_c_FI5RYWbdpAtjsy4OIKrKw'
const tldrawTestThreePagesUrl = 'https://www.tldraw.com/s/v2_c_L_RFQ3mJA_BWHejdH2hlD'

it(
	'should export the tldraw url to an svg in the current folder by default',
	{ timeout: 10_000 },
	async () => {
		const [savedImageFileName] = await tldrawToImage(tldrawTestUrl)

		expect(savedImageFileName).toBe(`${process.cwd()}/v2_c_9nMYBwT8UQ99RGDWfGr8H.svg`)
		expectFileToBeValid(savedImageFileName, 'svg')

		if (cleanUp) rmSync(savedImageFileName)
	},
)

it('should export the tldraw url to an svg when specified', { timeout: 10_000 }, async () => {
	const [savedImageFileName] = await tldrawToImage(tldrawTestUrl, { format: 'svg' })

	expect(savedImageFileName).toBe(`${process.cwd()}/v2_c_9nMYBwT8UQ99RGDWfGr8H.svg`)
	expectFileToBeValid(savedImageFileName, 'svg')

	if (cleanUp) rmSync(savedImageFileName)
})

it('should export the tldraw url to a png when specified', { timeout: 10_000 }, async () => {
	const [savedImageFileName] = await tldrawToImage(tldrawTestUrl, { format: 'png' })

	expect(savedImageFileName).toBe(`${process.cwd()}/v2_c_9nMYBwT8UQ99RGDWfGr8H.png`)
	expectFileToBeValid(savedImageFileName, 'png')

	if (cleanUp) rmSync(savedImageFileName)
})

it(
	'should export the file to a specific directory when specified',
	{ timeout: 10_000 },
	async () => {
		const randomPath = nanoid()
		mkdirSync(randomPath)

		const [savedImageFileName] = await tldrawToImage(tldrawTestUrl, {
			format: 'png',
			output: `./${randomPath}/`,
		})
		expect(savedImageFileName).toContain(randomPath)

		expectFileToBeValid(savedImageFileName, 'png')

		if (cleanUp) rmSync(randomPath, { recursive: true })
	},
)

it(
	'should export the entire image if multiple frames are present and --frames is not set',
	{ timeout: 10_000 },
	async () => {
		const [savedImageFileName] = await tldrawToImage(tldrawTestThreeFramesUrl)

		expectFileToBeValid(savedImageFileName, 'svg')

		if (cleanUp) rmSync(savedImageFileName)
	},
)

it('should export each frame individually if --frames is set', { timeout: 10_000 }, async () => {
	const savedImageFileNames = await tldrawToImage(tldrawTestThreeFramesUrl, {
		frames: true,
	})

	expect(savedImageFileNames).toHaveLength(3)

	for (const fileName of savedImageFileNames) {
		expectFileToBeValid(fileName, 'svg')
	}

	if (cleanUp) {
		for (const fileName of savedImageFileNames) {
			rmSync(fileName)
		}
	}
})

it('should export each page individually if --pages is set', { timeout: 10_000 }, async () => {
	const savedImageFileNames = await tldrawToImage(tldrawTestThreePagesUrl, {
		pages: true,
	})

	expect(savedImageFileNames).toHaveLength(3)

	for (const fileName of savedImageFileNames) {
		expectFileToBeValid(fileName, 'svg')
	}

	if (cleanUp) {
		for (const fileName of savedImageFileNames) {
			rmSync(fileName)
		}
	}
})

it('should export to json', { timeout: 10_000 }, async () => {
	const [savedImageFileName] = await tldrawToImage(tldrawTestUrl, {
		format: 'json',
	})

	expect(savedImageFileName).toBe(`${process.cwd()}/v2_c_9nMYBwT8UQ99RGDWfGr8H.json`)
	expectFileToBeValid(savedImageFileName, 'json')

	if (cleanUp) rmSync(savedImageFileName)
})

it('should export frames to json', { timeout: 10_000 }, async () => {
	const savedImageFileNames = await tldrawToImage(tldrawTestThreeFramesUrl, {
		format: 'json',
		frames: true,
	})

	expect(savedImageFileNames).toHaveLength(3)

	for (const fileName of savedImageFileNames) {
		expectFileToBeValid(fileName, 'json')
	}

	if (cleanUp) {
		for (const fileName of savedImageFileNames) {
			rmSync(fileName)
		}
	}
})

it('should export to tldr', { timeout: 10_000 }, async () => {
	const [savedImageFileName] = await tldrawToImage(tldrawTestUrl, {
		format: 'tldr',
	})

	expect(savedImageFileName).toBe(`${process.cwd()}/v2_c_9nMYBwT8UQ99RGDWfGr8H.tldr`)
	expectFileToBeValid(savedImageFileName, 'tldr')

	if (cleanUp) rmSync(savedImageFileName)
})

it('should export schema 2 sketches to svg', { timeout: 10_000 }, async () => {
	const [savedImageFileName] = await tldrawToImage(tldrawTestUrlSchema2)

	expect(savedImageFileName).toBe(`${process.cwd()}/v2_c_85CiFqkLgUaiwmed4kIa_.svg`)
	expectFileToBeValid(savedImageFileName, 'svg')

	if (cleanUp) rmSync(savedImageFileName)
})

it('should export schema 2 sketches to tldr', { timeout: 10_000 }, async () => {
	const [savedImageFileName] = await tldrawToImage(tldrawTestUrlSchema2, {
		format: 'tldr',
	})

	expect(savedImageFileName).toBe(`${process.cwd()}/v2_c_85CiFqkLgUaiwmed4kIa_.tldr`)
	expectFileToBeValid(savedImageFileName, 'tldr')

	if (cleanUp) rmSync(savedImageFileName)
})
