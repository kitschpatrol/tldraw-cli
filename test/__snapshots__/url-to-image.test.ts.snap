// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should export each frame individually if --frames is set 1`] = `"7918cb9fbb25056d47f5b9b9ac4831d3cc95a93494d1558944b310ca72062021"`;

exports[`should export each frame individually if --frames is set 2`] = `"9e869454835720d439ea4a04d0e792e3bce1bf6ad8a038527ff31af8cb82ce01"`;

exports[`should export each frame individually if --frames is set 3`] = `"5cb3eb0b9f4dcf7ed10dfdbd00e7df31750ebfe47a4726dfc2910d00635fb0f9"`;

exports[`should export each page individually if --pages is set 1`] = `"a62aaa4e6730a20162f9ae40f0c93e4cc9ee04770129ee6ec64edc2536fbed3d"`;

exports[`should export each page individually if --pages is set 2`] = `"cb0a78e0dbcb6152c6a883bded4ae20d4dd07c45a70d9a460c2d293db9d12a51"`;

exports[`should export each page individually if --pages is set 3`] = `"e7907f1cf15ac7c05427ae2b0b5256bad1d13d0e29c01b46571bdfe0ecf7a21d"`;

exports[`should export frames to json 1`] = `"042730a1f892f91f5bf18cf9bdaada3ce484fdfd349221f7345346c1e4d02550"`;

exports[`should export frames to json 2`] = `"5cdb89ab84843756deb0b7c50c61dc1db6277a1011f4fe33b6f6055805d67187"`;

exports[`should export frames to json 3`] = `"ca4d8de3c424cdd4c50a0f4fadbfa57882a94cf5966bad733edf21a214d80bb9"`;

exports[`should export schema 2 sketches to svg 1`] = `"e478ab790d8ec5f216f7e9b3a86b183455e604ad5646b748e5a9d929db05a0ca"`;

exports[`should export schema 2 sketches to tldr 1`] = `"8eff051fe82716ccc1bc7ce73eb35e6bfe0977d786f1ce4d5c840a104d926d9e"`;

exports[`should export the entire image if multiple frames are present and --frames is not set 1`] = `"f060c82ae2ca58cd0100e768415988e0eaa2e7706330949c0364ca5282a95700"`;

exports[`should export the file to a specific directory when specified 1`] = `"1100111001110010101100011011001110110010101100110100110100110011"`;

exports[`should export the tldraw url to a png when specified 1`] = `"1100111001110010101100011011001110110010101100110100110100110011"`;

exports[`should export the tldraw url to an svg in the current folder by default 1`] = `"2f1f7a5e9c8b14db98e227b470957510ec561fb1cfe571a64c55105f0a4a3f94"`;

exports[`should export the tldraw url to an svg when specified 1`] = `"2f1f7a5e9c8b14db98e227b470957510ec561fb1cfe571a64c55105f0a4a3f94"`;

exports[`should export to json 1`] = `"06bef173a8a2e1e2dabe83444b3758bfa5afa83e1862dd51eafcfc5d67852d38"`;

exports[`should export to tldr 1`] = `"de5f381a80db846dd0940a21c96aeff27e52155e86595bb8f9aae59fe2a15f23"`;
